### PR TITLE
[Storage][Sharding] Split state k/v commit from ledger db commit.

### PR DESCRIPTION
**Pull request by grao1991**
### Description
Commit state k/v and the rest of data in parallel in different Rocksdb commits. Add logic at the start to ensure they are consistent.

The change of using a separate database for state k/v is guarded by a config flag, thus without turning on the flag, this just split a commit batch into two batches, and both will still get committed into ledger db.

Pruner is not handled in this PR.
Truncation tool is not handled in this PR.

### CODE REVIEW

File: config/src/config/storage_config.rs [hunk 1]
```diff
@@ -63,8 +64,8 @@ impl Default for RocksdbConfigs {
         Self {
             ledger_db_config: RocksdbConfig::default(),
             state_merkle_db_config: RocksdbConfig::default(),
-            use_kv_db: false,
-            kv_db_config: RocksdbConfig::default(),
+            use_state_kv_db: false,
+            state_kv_db_config: RocksdbConfig::default(),
```
Comment (msmouse) [inline config/src/config/storage_config.rs:68]: love the rename, please make it a separate commit though

File: storage/aptosdb/src/state_store/mod.rs [hunk 2]
```diff
@@ -267,19 +274,77 @@ impl StateStore {
...
+                .expect("State K/V commit progress cannot be None.")
+                .expect_version();
+            assert_ge!(state_kv_commit_progress, overall_commit_progress);
+
+            if ledger_commit_progress != overall_commit_progress {
+                info!(
+                    "Ledger commit progress is {}, start truncation...",
+                    ledger_commit_progress
+                );
+                let difference = ledger_commit_progress - overall_commit_progress;
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/state_store/mod.rs:311]: I have assert_ge! above.

**Comment by sitalkedia** [comment]
@grao1991 - Hmm I am not seeing any measurable difference in commit latency. Compare http://o11y.aptosdev.com/grafana/goto/0EqJ3h1Vz?orgId=1 (current PR) with http://o11y.aptosdev.com/grafana/goto/GiIT3h14k?orgId=1 (without this). Is that expected?

**Comment by grao1991** [comment]
> @grao1991 - Hmm I am not seeing any measurable difference in commit latency. Compare http://o11y.aptosdev.com/grafana/goto/0EqJ3h1Vz?orgId=1 (current PR) with http://o11y.aptosdev.com/grafana/goto/GiIT3h14k?orgId=1 (without this). Is that expected?

Yes, the flag is not turned on. (not ready to turn it on on production since I don't consider the migration yet.)

**Comment by sitalkedia** [comment]
Let's turn on the flag in https://github.com/aptos-labs/aptos-core/pull/6634 in that case and we can compare the commit time with previous run.

File: storage/aptosdb/src/lib.rs [hunk 3]
```diff
@@ -1766,12 +1796,36 @@ impl DbWriter for AptosDB {
...
+                        let t0 = s.spawn(|| self.state_kv_db.write_schemas(state_kv_batch));
+                        let t1 = s.spawn(|| self.ledger_db.write_schemas(ledger_batch));
+                        t0.join().unwrap().unwrap();
+                        t1.join().unwrap().unwrap();
+                    });
+
+                    let ledger_batch = SchemaBatch::new();
+                    self.save_ledger_info(new_root_hash, ledger_info_with_sigs, &ledger_batch)?;
+                    ledger_batch.put::<DbMetadataSchema>(
+                        &DbMetadataKey::OverallCommitProgress,
```
Comment (areshand) [inline storage/aptosdb/src/lib.rs:1825]: curious about the motivation for introducing OverallCommitProgress?

File: storage/aptosdb/src/lib.rs [hunk 4]
```diff
@@ -265,15 +266,19 @@ impl AptosDB {
...
         max_nodes_per_lru_cache_shard: usize,
         hack_for_tests: bool,
     ) -> Self {
         let arc_ledger_rocksdb = Arc::new(ledger_rocksdb);
         let arc_state_merkle_rocksdb = Arc::new(state_merkle_rocksdb);
-        let arc_kv_rocksdb = Arc::new(kv_rocksdb);
+        let arc_state_kv_rocksdb = if let Some(db) = state_kv_rocksdb {
+            Arc::new(db)
+        } else {
+            Arc::clone(&arc_ledger_rocksdb)
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:280]: we won't access those CFs from state_kv_db.

File: storage/aptosdb/src/lib.rs [hunk 5]
```diff
@@ -399,21 +406,23 @@ impl AptosDB {
...
                 } else {
                     None
                 },
             )
         };
 
-        if rocksdb_configs.use_kv_db {
-            info!(kv_db_path = kv_db_path, "Opened K/V DB.",);
+        if rocksdb_configs.use_state_kv_db {
+            info!(state_kv_db_path = state_kv_db_path, "Opened state K/V DB.",);
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:423]: structural logging

File: storage/aptosdb/src/lib.rs [hunk 6]
```diff
@@ -501,13 +510,13 @@ impl AptosDB {
...
             )?,
-            if rocksdb_configs.use_kv_db {
+            if rocksdb_configs.use_state_kv_db {
                 Some(DB::open_cf_as_secondary(
-                    &gen_rocksdb_options(&rocksdb_configs.kv_db_config, false),
-                    kv_db_primary_path,
-                    kv_db_secondary_path,
-                    "kv_db_sec",
-                    kv_db_column_families(),
+                    &gen_rocksdb_options(&rocksdb_configs.state_kv_db_config, false),
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:515]: See https://github.com/facebook/rocksdb/wiki/Read-only-and-Secondary-instances

I don't think we currently use it because Alden observed some unexpected behaviors which we don't fully understand. But let's keep it in case we need it in the future.

File: storage/aptosdb/src/lib.rs (hunk 3 above)
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:1825]: It's easier to understand and maintain.
Also, one reason I don't use the LedgerInfo version as the progress here because we overwrite the previous LedgerInfo in the same epoch in every commit. If we use the db truncation tool to truncate the db, the ledger info version will not be the same as the target version of the truncation. In this case I'd rely on this OverallCommitProgress.

File: storage/aptosdb/src/state_store/mod.rs [hunk 7]
```diff
@@ -267,19 +274,77 @@ impl StateStore {
...
         hack_for_tests: bool,
     ) -> Self {
+        if let Some(DbMetadataValue::Version(overall_commit_progress)) = ledger_db
+            .get::<DbMetadataSchema>(&DbMetadataKey::OverallCommitProgress)
+            .expect("Failed to read overall commit progress.")
+        {
+            info!(
+                "Overall commit progress is {}, start syncing databases...",
+                overall_commit_progress
+            );
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/state_store/mod.rs:291]: Done

File: config/src/config/storage_config.rs (hunk 1 above)
Comment (grao1991) [reply config/src/config/storage_config.rs:68]: Done

File: storage/aptosdb/src/db_options.rs [hunk 8]
```diff
@@ -43,9 +43,10 @@ pub(super) fn state_merkle_db_column_families() -> Vec<ColumnFamilyName> {
     ]
 }
 
-pub(super) fn kv_db_column_families() -> Vec<ColumnFamilyName> {
+pub(super) fn state_kv_db_column_families() -> Vec<ColumnFamilyName> {
     vec![
         /* empty cf */ DEFAULT_COLUMN_FAMILY_NAME,
+        DB_METADATA_CF_NAME,
```
Comment (areshand) [inline storage/aptosdb/src/db_options.rs:49]: why do we want to have metadata and the data in the same CF?

File: execution/executor-benchmark/src/db_generator.rs [hunk 9]
```diff
@@ -25,6 +25,7 @@ pub fn run<V>(
     db_dir: impl AsRef<Path>,
     storage_pruner_config: PrunerConfig,
     verify_sequence_numbers: bool,
+    use_state_kv_db: bool,
```
Comment (areshand) [inline execution/executor-benchmark/src/db_generator.rs:28]: nit: if we have more split in the future, maybe consider passing a struct containing multiple flags so we don't need to change the function signature?

Comment (grao1991) [reply execution/executor-benchmark/src/db_generator.rs:28]: I haven't decided yet. If that's the case I will probably pass the StroageConfig directly.

**Review by lightmark — APPROVED** [review]

File: storage/aptosdb/src/db_options.rs (hunk 8 above)
Comment (areshand) [inline storage/aptosdb/src/db_options.rs:49]: why do we store DB metadata in every CF? Hav we consider a separate CF for metadata ?

Comment (grao1991) [reply storage/aptosdb/src/db_options.rs:49]: It is a separate CF.

File: storage/aptosdb/src/lib.rs (hunk 3 above)
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:1825]: https://github.com/aptos-labs/aptos-core/pull/6482

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/state_store/mod.rs [hunk 10]
```diff
@@ -267,19 +274,77 @@ impl StateStore {
...
         ledger_db: Arc<DB>,
         state_merkle_db: Arc<DB>,
+        state_kv_db: Arc<DB>,
         state_pruner: StatePrunerManager<StaleNodeIndexSchema>,
         epoch_snapshot_pruner: StatePrunerManager<StaleNodeIndexCrossEpochSchema>,
         buffered_state_target_items: usize,
         max_nodes_per_lru_cache_shard: usize,
         hack_for_tests: bool,
     ) -> Self {
+        if let Some(DbMetadataValue::Version(overall_commit_progress)) = ledger_db
```
Comment (ppoliani) [inline storage/aptosdb/src/state_store/mod.rs:284]: Would it make sense to disable the truncation part? At the moment it's making my full node fail https://github.com/aptos-labs/aptos-core/issues/7982 

Would this corrupt the state if just comment out this added code?

Comment (msmouse) [reply storage/aptosdb/src/state_store/mod.rs:284]: (addressed here https://github.com/aptos-labs/aptos-core/issues/7985)
